### PR TITLE
dump operator names of a module and its sub-modules.

### PR DESCRIPTION
**Pull request by iseeyuan**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29208 dump operator names of a module and its sub-modules.**

A binary to dump operator names from a script model and its sub models. 
Usage:
dump_operator_names path/to/script_model.pt path/to/output.yaml

Differential Revision: [D18350353](https://our.internmc.facebook.com/intern/diff/D18350353)

### CODE REVIEW

File: binaries/dump_operator_names.cc [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        std::string namestr = opname.name;
+        if (!opname.overload_name.empty())
+          namestr += "." + opname.overload_name;
+        std::cout << "    " << namestr << std::endl;
+        opnames.emplace(namestr);
+      }
+    }
+  }
+  auto modules = m.get_modules();
+  for (const auto& sub_m : modules) {
```
Comment (ljk53) [inline binaries/dump_operator_names.cc]: Can sub-module have sub-module too?

File: binaries/dump_operator_names.cc [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+}
+}
+}
+
+int main(int argc, char** argv) {
+  if (argc != 3) {
+    std::cerr << "usage: <path-to-script-module> <path-to-output-yaml>\n";
+    return 1;
+  }
+  auto m = torch::jit::load(argv[1]);
```
Comment (ljk53) [inline binaries/dump_operator_names.cc]: nit: can you use C10_DEFINE_string to define & parse command line flags, like other binaries/*.cpp files?

File: binaries/dump_operator_names.cc [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+#include <torch/script.h>
+
+#include <fstream>
+
+namespace torch {
+namespace jit {
+void dump_opnames(const script::Module& m, std::unordered_set<std::string>& opnames) {
+  auto methods = m.get_methods();
+  for (const auto &method : methods) {
+    const auto &func = method.function();
```
Comment (ljk53) [inline binaries/dump_operator_names.cc]: nit: "const auto &" -> "const auto&", to be consistent with other places in this file.

File: binaries/dump_operator_names.cc (hunk 1 above)
Comment (iseeyuan) [reply binaries/dump_operator_names.cc]: Yes. It's a recursive call (line 35) and should cover all levels of submodules.

File: binaries/dump_operator_names.cc (hunk 3 above)
Comment (ljk53) [reply binaries/dump_operator_names.cc]: should update both line 16 and 17?

File: binaries/dump_operator_names.cc [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    std::cout << "function name: " << func.name() << std::endl;
+    torch::jit::Code code(func.graph());
+    for (size_t i = 0; i < code.instructions().size(); ++i) {
+      auto ins = code.instructions()[i];
+      auto node = code.instructions_source()[i];
+      if (ins.op == OpCode::OP) {
+        auto opname = node->schema().operator_name();
+        std::string namestr = opname.name;
+        if (!opname.overload_name.empty())
+          namestr += "." + opname.overload_name;
```
Comment (ljk53) [inline binaries/dump_operator_names.cc]: is overload_name being used? the codegen script will take the output of this binary as a whitelist of root ops that need to be kept. If it needs check both opname and overload_name then probably it's easier to write them as separate yaml fields? unless opname and overload_name are guaranteed to be concatenated with "." in the codegen script as well.
we could do this in separate PR if needed.

**Review by ljk53 — APPROVED** [review]

File: binaries/dump_operator_names.cc [hunk 5]
```diff
@@ -0,0 +1,69 @@
+#include <torch/csrc/jit/script/module.h>
```
Comment (ljk53) [inline binaries/dump_operator_names.cc]: Note that majority of files under binaries/ have "Copyright (c) 2016-present, Facebook, Inc." license header. Consider following this convention?

File: binaries/dump_operator_names.cc (hunk 4 above)
Comment (iseeyuan) [reply binaries/dump_operator_names.cc]: I think that opname and overload_name are planed to be concatenated with "." in the codegen script. Most of aten ops are already like that, except some jit prim/special ops that miss the overload_name field, but they are going to be updated to "opname.overload_name" format. Is it possible to separate the fields when parsing the yaml file?

Comment (iseeyuan) [reply binaries/dump_operator_names.cc]: @smessmer Could you confirm the name.overload_name format?

**Comment by pietern** [comment]
@iseeyuan This raced with #28828 and broke master.

Working on a fix instead of reverting because I think it's an easy fix.

**Comment by pietern** [comment]
Turns out the fix is not as quick as I wanted. Reverting this PR.
